app/scripts/controllers/transactions/index.js

```diff
@@ -389,11 +391,17 @@ class TransactionController extends EventEmitter {
     // sign tx
     const fromAddress = txParams.from
     const ethTx = new Transaction(txParams)
+    ethTx.r = ethUtil.toBuffer(txId)
```

Comment:
sorry, realized you are setting the `r` signature value on ethereumjs-tx to the...`txId`? why?

Reply:
I see, that makes sense. I think using the keyring to trigger the qr code reader is not a good idea in this case. I think a better way to approach this would be:
- the tx confirmation flow checks if the keyring type for the account has a special ui signing flow
- if it does that flow is routed to
- BidirectionalQrCode signers use this flow to perform and relay the signature

Reply:
I actually started to do things like you suggested: since it is a client side thing to display the QR data to user. And actually It is a way less of programming and figuring out the internal structure of MetaMask.
The reason why I scraped this solution is that the nonce is calculated AFTER the user has hit CONFIRM button, this means that you can not display a tx QR code to user, since nonce is not known. Actually it is created only just before the tx gets actually signed by eth-keyring-controller. The other reason I did not like the UI side solution is that you have at least five UIs where you have to mess with the code: ETH transfer, token transfer, contract execution, message signing, typed data signing etc. And of course for all of these you have to create a custom code that calls the background in a special way with tx already signed.

It is a much cleaner solution to use keyrings as all the signing end up there anyway no matter if it is token tx, eth tx, contract tx, or whatever else. It is also futureproof.

And I also have pretty bad feelings about your comment since I've been doing this work on my free time, I did it because I am using bidirectional QR to interact with crypto and it seemed to be fun to do it with Metamask. I have put quite lot of work into this, but to rewrite the whole code again.... just seems to be too much for me, so I see two bad choices either I rewrite it and put a lot of effort again, or abandon the code.

So let's see if we can figure out something about this!

Reply:
BTW found a better soulution and removed `ethTx.r = ethUtil.toBuffer(txId)` above.

Reply:
Is there any good ways with which a Keyring can expose some of it's state to Metamask UIs? That would make the modification of `eth-keyring-controller` unnecessary.

Reply:
> Longer explanation: When the background notifies all the UI s that a QR signing modal should be opened, we want only the one UI to render the modal only where the "Confirm" button was hit.

I have a hard time imagining a way to manage the chaos here. Of course the problem can be solved but I think it will always be weird and squishy. What happens when you close the UI? can you find it again? Completing the signature via QR immediately after approval feels like the most manageable, sane thing to me.

> The other reason I did not like the UI side solution is that you have at least five UIs where you have to mess with the code: ETH transfer, token transfer, contract execution, message signing, typed data signing etc.

If done correctly, we should only need one. flow like:
1. relevant signature confirm view (eth, token, ...)
2. keyring specific signing view (just loading spinner, qr code)

>  the nonce is calculated AFTER the user has hit CONFIRM button, this means that you can not display a tx QR code to user, since nonce is not known.

yeah thats a good point. we would need to expose the nonce-mutex to the signing ui somehow.

> Is there any good ways with which a Keyring can expose some of it's state to Metamask UIs?

yeah that shouldnt be too hard, sounds like a good idea. I think this may make everything much easier to do. both the signing as part of the UI flow AND using the keyring. The keyring could hold the state of the signing flow and the UI could use that state to display the QR code. Still does not cover the case where the ui is closed during the signature flow.

Reply:
> Still does not cover the case where the ui is closed during the signature flow.

It is done. The UI sends keep alive to BG in every 5 secs. If no more send  happens tx is considered canceled. The code should be operational now.